### ISSUE TITLE
__main__.py uses global youtube_dl package

### ISSUE THREAD

**Issue by emphoeller** [ytdl-org/youtube-dl#30855]
## Checklist

- [x] I'm reporting a broken site support
- [x] I've verified that I'm running youtube-dl version **2021.12.17**
- [x] I've checked that all provided URLs are alive and playable in a browser
- [x] I've checked that all URLs and arguments with special characters are properly quoted or escaped
- [x] I've searched the bugtracker for similar issues including closed ones

Note: This *is* using [release 2021.12.17](https://github.com/ytdl-org/youtube-dl/releases/tag/2021.12.17), but `--version` still prints `2021.06.06`.

## Verbose log

```
python3 youtube_dl-2021.12.17/youtube_dl -v 'https://www.youtube.com/watch?v=ZPpKrG1Pp5M'
[debug] System config: []
[debug] User config: []
[debug] Custom config: []
[debug] Command-line args: ['-v', 'https://www.youtube.com/watch?v=ZPpKrG1Pp5M']
[debug] Encodings: locale UTF-8, fs utf-8, out UTF-8, pref UTF-8
[debug] youtube-dl version 2021.06.06
[debug] Python version 3.6.5 (CPython) - Linux-5.15.15-pclos1-x86_64-with-mandrake-2022-PCLinuxOS
[debug] exe versions: ffmpeg 4.4.1, ffprobe 4.4.1
[debug] Proxy map: {}
[youtube] ZPpKrG1Pp5M: Downloading webpage
[youtube] ZPpKrG1Pp5M: Downloading player 0c665041
ERROR: Signature extraction failed: Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/youtube_dl/extractor/youtube.py", line 1349, in _decrypt_signature
    video_id, player_url, s
  File "/usr/lib/python3.6/site-packages/youtube_dl/extractor/youtube.py", line 1265, in _extract_signature_function
    cache_res = res(test_string)
  File "/usr/lib/python3.6/site-packages/youtube_dl/extractor/youtube.py", line 1332, in <lambda>
    return lambda s: initial_function([s])
  File "/usr/lib/python3.6/site-packages/youtube_dl/jsinterp.py", line 258, in resf
    res, abort = self.interpret_statement(stmt, local_vars)
  File "/usr/lib/python3.6/site-packages/youtube_dl/jsinterp.py", line 56, in interpret_statement
    v = self.interpret_expression(expr, local_vars, allow_recursion)
  File "/usr/lib/python3.6/site-packages/youtube_dl/jsinterp.py", line 150, in interpret_expression
    return obj[member]
TypeError: string indices must be integers
 (caused by TypeError('string indices must be integers',)); please report this issue on https://yt-dl.org/bug . Make sure you are using the latest version; see  https://yt-dl.org/update  on how to update. Be sure to call youtube-dl with the --verbose flag and include its complete output.
Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/youtube_dl/extractor/youtube.py", line 1349, in _decrypt_signature
    video_id, player_url, s
  File "/usr/lib/python3.6/site-packages/youtube_dl/extractor/youtube.py", line 1265, in _extract_signature_function
    cache_res = res(test_string)
  File "/usr/lib/python3.6/site-packages/youtube_dl/extractor/youtube.py", line 1332, in <lambda>
    return lambda s: initial_function([s])
  File "/usr/lib/python3.6/site-packages/youtube_dl/jsinterp.py", line 258, in resf
    res, abort = self.interpret_statement(stmt, local_vars)
  File "/usr/lib/python3.6/site-packages/youtube_dl/jsinterp.py", line 56, in interpret_statement
    v = self.interpret_expression(expr, local_vars, allow_recursion)
  File "/usr/lib/python3.6/site-packages/youtube_dl/jsinterp.py", line 150, in interpret_expression
    return obj[member]
TypeError: string indices must be integers
Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/youtube_dl/extractor/youtube.py", line 1349, in _decrypt_signature
    video_id, player_url, s
  File "/usr/lib/python3.6/site-packages/youtube_dl/extractor/youtube.py", line 1265, in _extract_signature_function
    cache_res = res(test_string)
  File "/usr/lib/python3.6/site-packages/youtube_dl/extractor/youtube.py", line 1332, in <lambda>
    return lambda s: initial_function([s])
  File "/usr/lib/python3.6/site-packages/youtube_dl/jsinterp.py", line 258, in resf
    res, abort = self.interpret_statement(stmt, local_vars)
  File "/usr/lib/python3.6/site-packages/youtube_dl/jsinterp.py", line 56, in interpret_statement
    v = self.interpret_expression(expr, local_vars, allow_recursion)
  File "/usr/lib/python3.6/site-packages/youtube_dl/jsinterp.py", line 150, in interpret_expression
    return obj[member]
TypeError: string indices must be integers

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/youtube_dl/YoutubeDL.py", line 815, in wrapper
    return func(self, *args, **kwargs)
  File "/usr/lib/python3.6/site-packages/youtube_dl/YoutubeDL.py", line 836, in __extract_info
    ie_result = ie.extract(url)
  File "/usr/lib/python3.6/site-packages/youtube_dl/extractor/common.py", line 534, in extract
    ie_result = self._real_extract(url)
  File "/usr/lib/python3.6/site-packages/youtube_dl/extractor/youtube.py", line 1623, in _real_extract
    signature = self._decrypt_signature(sc['s'][0], video_id, player_url)
  File "/usr/lib/python3.6/site-packages/youtube_dl/extractor/youtube.py", line 1359, in _decrypt_signature
    'Signature extraction failed: ' + tb, cause=e)
youtube_dl.utils.ExtractorError: Signature extraction failed: Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/youtube_dl/extractor/youtube.py", line 1349, in _decrypt_signature
    video_id, player_url, s
  File "/usr/lib/python3.6/site-packages/youtube_dl/extractor/youtube.py", line 1265, in _extract_signature_function
    cache_res = res(test_string)
  File "/usr/lib/python3.6/site-packages/youtube_dl/extractor/youtube.py", line 1332, in <lambda>
    return lambda s: initial_function([s])
  File "/usr/lib/python3.6/site-packages/youtube_dl/jsinterp.py", line 258, in resf
    res, abort = self.interpret_statement(stmt, local_vars)
  File "/usr/lib/python3.6/site-packages/youtube_dl/jsinterp.py", line 56, in interpret_statement
    v = self.interpret_expression(expr, local_vars, allow_recursion)
  File "/usr/lib/python3.6/site-packages/youtube_dl/jsinterp.py", line 150, in interpret_expression
    return obj[member]
TypeError: string indices must be integers
 (caused by TypeError('string indices must be integers',)); please report this issue on https://yt-dl.org/bug . Make sure you are using the latest version; see  https://yt-dl.org/update  on how to update. Be sure to call youtube-dl with the --verbose flag and include its complete output.

```

## Description

When attempting to download that paticular video, I get the above error. Other videos (e.g. `'https://www.youtube.com/watch?v=SwUBOyrxioI'`) still work for me.

**Comment by emphoeller** [comment]
I just noticed that this seems to be using the version of youtube-dl I installed with pip some time ago, as the files listed are in my python package directory, even though I explicitly invoked a version of the package located elsewhere. Apparently, the `import youtube_dl` in `__main__.py` finds the directory `youtube_dl` it is contained in only if the current directory is the parent directory of `youtube_dl`; otherwise the import silently falls back to the globally installed package. The download works with the correct current directory. Closing.

**Comment by emphoeller** [comment]
Nope, it wasn’t the current directory (that doesn’t change anything), it was invoking `youtube-dl` instead of `youtube_dl`. `import` always searches relative to the file that the statement appears in, so `__main__.py` couldn’t possibly find the local `youtube_dl`. This is a bug.

**Comment by 89z** [comment]
Seems to work fine:

~~~
PS D:\Desktop> pip install youtube-dl
PS D:\Desktop> youtube-dl.exe ZPpKrG1Pp5M
[youtube] ZPpKrG1Pp5M: Downloading webpage
[youtube] ZPpKrG1Pp5M: Downloading player 0c665041
WARNING: Requested formats are incompatible for merge and will be merged into mkv.
[download] Destination: BC3!!!! 2-ZPpKrG1Pp5M.f137.mp4
[download]   3.5% of 19.80MiB at 73.48KiB/s ETA 04:26

PS D:\Desktop> pip install https://github.com/ytdl-org/youtube-dl/archive/refs/heads/master.zip
PS D:\Desktop> youtube-dl.exe ZPpKrG1Pp5M
[youtube] ZPpKrG1Pp5M: Downloading webpage
WARNING: Requested formats are incompatible for merge and will be merged into mkv.
[download] Resuming download at byte 733079
[download] Destination: BC3!!!! 2-ZPpKrG1Pp5M.f137.mp4
[download]  72.5% of 19.80MiB at  8.26MiB/s ETA 00:00
~~~

**Comment by emphoeller** [comment]
There is no problem if you are using the precompiled `youtube-dl` or the standalone `youtube-dl.exe`. But if you have version A installed via pip and then try to run version B by calling its `youtube_dl` directory (which will invoke `__main__.py`), you will get version A instead of the version B you are expecting.

**Comment by emphoeller** [comment]
Even if the version installed via pip is the most recent one, a developer might want to test a change by running `python3 youtube_dl` in their working directory. This will opaquely run the version installed via pip, however.

**Comment by dirkf** [comment]
Use `-m path/to/desired/youtube_dl`.

**Comment by emphoeller** [comment]
It still doesn’t seem like intended (or tolerable) behavior to me that omitting the `-m` will just run the global module opaquely. The least we could do is exit with an error and tell the user how to do it correctly. I am so dissatisfied with this because it has already bitten me some time ago when I tried to contribute to this project.

Luckily, there is an easy fix:

```python
if __package__ is None and not hasattr(sys, 'frozen'):
    # direct call of __main__.py
```

in `__main__.py` should become

```python
if not __package__ and not hasattr(sys, 'frozen'):
    # call of the youtube_dl dir ('') or direct call of __main__.py (None)
```

For some reason, there is code (under the `if`) that fixes this for `python youtube_dl/__main__.py` (in which case `__package__` is `None`) by putting the directory that contains the `youtube_dl` directory at the beginning of `sys.path` but that is written in such a way that it won’t do the same thing for `python youtube_dl` (in which case `__package__` is `''`). The above change enables the fix for the second case.

So how does this work? Do I create a pull request for two lines of code?

**Comment by dirkf** [comment]
I believe the current behaviour is by design (https://github.com/ytdl-org/youtube-dl/commit/743b28ce1155df348bbb828fcb316d57c8704328)
```
# Execute with
# $ python youtube_dl/__main__.py (2.6+)
# $ python -m youtube_dl          (2.7+)
...
```
As the change from `__package__is not None` to `not __package__` is very small, I'd be happy to make it once we're sure it works on all the supported platforms.

**Comment by emphoeller** [comment]
> I believe the current behaviour is by design

In that case it would be very poor design: As you pointed out, `__main__.py` will be run if its containing directory is passed to the Python interpreter. So a programmer will naturally assume that invoking the program by passing the directory is intended. But turns out it isn’t – and doing so might not fail, but silently run a version of youtube-dl installed elsewhere. Had the name simply been `main.py`, passing the directory would have led to an error, and there would have been no reason to assume that that was an option in the first place.

I just came back here to vent about how strange and misleading this is. It really frustrated me that the name makes no sense and I needed to point it out.
